pkg/iter/entry_iterator.go

```diff
@@ -344,28 +342,61 @@ func NewSortEntryIterator(is []EntryIterator, direction logproto.Direction) Entr
 	if len(is) == 1 {
 		return is[0]
 	}
-	result := &entrySortIterator{is: is}
-	switch direction {
-	case logproto.BACKWARD:
-		result.heap = &iteratorMaxHeap{iteratorHeap: make([]EntryIterator, 0, len(is))}
-	case logproto.FORWARD:
-		result.heap = &iteratorMinHeap{iteratorHeap: make([]EntryIterator, 0, len(is))}
-	default:
-		panic("bad direction")
-	}
+
+	result := &entrySortIterator{is: is, dir: direction}
 	return result
 }
 
-// init initialize the underlaying heap
+func (it *entrySortIterator) less(i, j int) bool {
+	t1, t2 := it.is[i].Entry().Timestamp, it.is[j].Entry().Timestamp
+	if it.dir == logproto.BACKWARD {
+		t2, t1 = t1, t2
+
+	}
+
+	un1 := t1.UnixNano()
+	un2 := t2.UnixNano()
+
+	switch {
+	case un1 < un2:
+		return true
+	case un1 > un2:
+		return false
+	default: // un1 == un2:
+		return it.is[i].Labels() < it.is[j].Labels()
+	}
+}
+
+func (it *entrySortIterator) lessThan(t1 time.Time, l1 string, j int) bool {
+	t2 := it.is[j].Entry().Timestamp
+	if it.dir == logproto.BACKWARD {
+		t2, t1 = t1, t2
+	}
```

Comment:
This is nasty. Ideally we would define the `lessThan` function once at the beginning.